src/component/mod.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2021-2022 Aaron Erhardt <aaron.erhardt@t-online.de>
+// Copyright 2022 System76 <info@system76.com>
+// SPDX-License-Identifier: MIT or Apache-2.0
+
+mod finalized;
+mod handle;
+
+#[allow(unreachable_pub)]
+pub use self::finalized::Finalized;
+#[allow(unreachable_pub)]
+pub use self::handle::Handle;
+
+use crate::{Receiver, Sender};
+use gtk::prelude::*;
+use tokio::sync::mpsc;
+
+#[async_trait::async_trait]
+/// Elm-style variant of a Component with view updates separated from input updates
+pub trait Component: Sized + 'static {
+    /// Internal commands to perform
+    type Command: 'static + Send;
+
+    /// The arguments that are passed to the init_view method.
+    type InitParams;
+
+    /// The message type that the component accepts as inputs.
+    type Input: 'static + Send;
+
+    /// The message type that the component provides as outputs.
+    type Output: 'static;
+
+    /// The widget that was constructed by the component.
+    type Root: Clone + AsRef<gtk::Widget>;
+
+    /// The type that's used for storing widgets created for this component.
+    type Widgets: 'static;
+
+    /// Initializes the root widget
+    fn init_root() -> Self::Root;
+
+    /// Initializes the model and view using the initial parameters.
+    fn init_inner(
+        params: Self::InitParams,
+        root: &Self::Root,
+        input: &mut Sender<Self::Input>,
+        output: &mut Sender<Self::Output>,
+    ) -> ComponentParts<Self, Self::Widgets>;
+
+    /// Initializes the component and attaches it to the thread-local GLib executor.
+    ///
+    /// The context's service remains alive for as long as the root widget is alive.
+    fn init(
```

Comment:
My idea for init was that it is already the second step. Let's think about the following example:

There's a component that needs to be added to your widgets in the correct order (between two other widgets). This component also need to call `set_transient_for` and therefore has the parent window in the `InitParams`. The problem now is that you have to pass the parent window before you have created the component. This requires some awkward manual ordering to create the window first, pass it to the component and then initialize the window with all its children.

My suggestion would be to just use `init_root()` for `RawComponent` and then use finalize for the actual initialization. With this the widget of a component is available from the beginning and all the ordering issues should disappear.